File: backend/app/schemas/tenant/announcement.py
"""
公告管理相关 Schema / Announcement management schemas.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import Field

from app.core.base_schema import (
    BaseCreateSchema,
    BaseResponseSchema,
    BaseSchema,
    BaseUpdateSchema,
    TenantResponseSchema,
)
from app.core.i18n import _

AnnouncementScope = Literal["admin", "tenant"]
AnnouncementStatus = Literal["draft", "published"]
AnnouncementPriority = Literal["low", "normal", "high", "urgent"]
AnnouncementRecipientType = Literal["admin", "tenant_admin", "tenant_user"]
AnnouncementFieldType = Literal["consent", "text", "radio", "checkbox"]


class AnnouncementFormOption(BaseSchema):
    """表单选项 / Form option."""

    label: str = Field(..., min_length=1)
    value: str = Field(..., min_length=1)


class AnnouncementFormField(BaseSchema):
    """公告反馈表单字段 / Announcement feedback form field."""

    key: str = Field(..., min_length=1)
    type: AnnouncementFieldType
    label: str = Field(..., min_length=1)
    required: bool = False
    placeholder: str | None = None
    options: list[AnnouncementFormOption] | None = None
    must_be_true: bool | None = None


class AnnouncementCreate(BaseCreateSchema):
    """创建公告草稿请求 / Create announcement draft request."""

    title: str = Field(..., min_length=1, description=_("tenant.announcement.field.title"))
    content: str | None = Field(None, description=_("tenant.announcement.field.content"))
    priority: AnnouncementPriority = Field(
        "normal", description=_("tenant.announcement.field.priority")
    )
    require_response: bool = Field(
        False, description=_("tenant.announcement.field.require_response")
    )
    form_schema: list[AnnouncementFormField] = Field(
        default_factory=list,
        description=_("tenant.announcement.field.form_schema"),
    )
    sort_order: int = Field(0, description=_("tenant.announcement.field.sort_order"))


class AnnouncementUpdate(BaseUpdateSchema):
    """更新公告草稿请求 / Update announcement draft request."""

    title: str | None = Field(None, min_length=1, description=_("tenant.announcement.field.title"))
    content: str | None = Field(None, description=_("tenant.announcement.field.content"))
    priority: AnnouncementPriority | None = Field(
        None, description=_("tenant.announcement.field.priority")
    )
    require_response: bool | None = Field(
        None, description=_("tenant.announcement.field.require_response")
    )
    form_schema: list[AnnouncementFormField] | None = Field(
        None,
        description=_("tenant.announcement.field.form_schema"),
    )
    sort_order: int | None = Field(
        None, description=_("tenant.announcement.field.sort_order")
    )


class AnnouncementAnswerSubmit(BaseSchema):
    """提交公告反馈 / Submit announcement feedback."""

    answers: dict[str, Any] = Field(default_factory=dict)


class AnnouncementBaseResponse(TenantResponseSchema):
    """公告基础响应 / Announcement base response."""

    scope: str = Field(..., description=_("tenant.announcement.field.scope"))
    title: str = Field(..., description=_("tenant.announcement.field.title"))
    content: str | None = Field(None, description=_("tenant.announcement.field.content"))
    status: str = Field(..., description=_("tenant.announcement.field.status"))
    priority: str = Field(..., description=_("tenant.announcement.field.priority"))
    require_response: bool = Field(
        ..., description=_("tenant.announcement.field.require_response")
    )
    form_schema: list[dict] = Field(
        default_factory=list,
        description=_("tenant.announcement.field.form_schema"),
    )
    published_at: datetime | None = Field(
        None, description=_("tenant.announcement.field.published_at")
    )
    recipient_count: int = Field(
        ..., description=_("tenant.announcement.field.recipient_count")
    )
    response_count: int = Field(
        ..., description=_("tenant.announcement.field.response_count")
    )
    sort_order: int = Field(..., description=_("tenant.announcement.field.sort_order"))

    @classmethod
    def from_model(cls, obj) -> AnnouncementBaseResponse:
        """从模型创建响应 / Build response from model."""
        return cls(
            id=obj.id,
            tenant_id=obj.tenant_id,
            scope=getattr(obj, "scope", None),
            title=getattr(obj, "title", None),
            content=getattr(obj, "content", None),
            status=getattr(obj, "status", None),
            priority=getattr(obj, "priority", None),
            require_response=getattr(obj, "require_response", False),
            form_schema=getattr(obj, "form_schema", None) or [],
            published_at=getattr(obj, "published_at", None),
            recipient_count=getattr(obj, "recipient_count", 0),
            response_count=getattr(obj, "response_count", 0),
            sort_order=getattr(obj, "sort_order", 0),
            created_at=obj.created_at,
            updated_at=obj.updated_at,
        )


class AdminAnnouncementResponse(AnnouncementBaseResponse):
    """公告管理响应（管理端）/ Announcement response (admin)."""


class TenantAnnouncementResponse(AnnouncementBaseResponse):
    """公告管理响应（企业端）/ Announcement response (tenant)."""


class AnnouncementDeliveryResponse(TenantResponseSchema):
    """公告投递响应 / Announcement delivery response."""

    announcement_id: int
    recipient_type: str
    recipient_id: int
    status: str
    notification_id: int | None = None
    submitted_at: datetime | None = None
    answers: dict[str, Any] | None = None

    @classmethod
    def from_delivery(cls, obj) -> AnnouncementDeliveryResponse:
        response = getattr(obj, "response", None)
        return cls(
            id=obj.id,
            tenant_id=obj.tenant_id,
            announcement_id=obj.announcement_id,
            recipient_type=obj.recipient_type,
            recipient_id=obj.recipient_id,
            status=obj.status,
            notification_id=obj.notification_id,
            submitted_at=obj.submitted_at,
            answers=getattr(response, "answers", None) if response else None,
            created_at=obj.created_at,
            updated_at=obj.updated_at,
        )


class PendingAnnouncementResponse(AnnouncementBaseResponse):
    """待处理公告响应 / Pending announcement response."""

    delivery_id: int

    @classmethod
    def from_delivery(cls, delivery) -> PendingAnnouncementResponse:
        announcement = delivery.announcement
        payload = AnnouncementBaseResponse.from_model(announcement).model_dump()
        payload["delivery_id"] = delivery.id
        return cls(**payload)


class CurrentAnnouncementResponse(AnnouncementBaseResponse):
    """当前接收人的公告详情 / Announcement detail for current recipient."""

    delivery_id: int
    delivery_status: str
    notification_id: int | None = None
    submitted_at: datetime | None = None
    answers: dict[str, Any] | None = None

    @classmethod
    def from_delivery(cls, delivery) -> CurrentAnnouncementResponse:
        announcement = delivery.announcement
        response = getattr(delivery, "response", None)
        payload = AnnouncementBaseResponse.from_model(announcement).model_dump()
        payload["delivery_id"] = delivery.id
        payload["delivery_status"] = delivery.status
        payload["notification_id"] = delivery.notification_id
        payload["submitted_at"] = delivery.submitted_at
        payload["answers"] = getattr(response, "answers", None) if response else None
        return cls(**payload)


class AnnouncementSubmitResult(BaseResponseSchema):
    """公告反馈提交结果 / Announcement feedback submit result."""

    announcement_id: int
    delivery_id: int
    status: str

    @classmethod
    def from_delivery(cls, delivery) -> AnnouncementSubmitResult:
        return cls(
            id=delivery.id,
            announcement_id=delivery.announcement_id,
            delivery_id=delivery.id,
            status=delivery.status,
            created_at=delivery.created_at,
            updated_at=delivery.updated_at,
        )


__all__ = [
    "AnnouncementAnswerSubmit",
    "AnnouncementCreate",
    "AnnouncementDeliveryResponse",
    "AnnouncementFormField",
    "AnnouncementFormOption",
    "AnnouncementSubmitResult",
    "AnnouncementUpdate",
    "AdminAnnouncementResponse",
    "CurrentAnnouncementResponse",
    "PendingAnnouncementResponse",
    "TenantAnnouncementResponse",
]

File: frontend/apps/web-antd/src/types/announcement.ts
export type AnnouncementFieldType = 'checkbox' | 'consent' | 'radio' | 'text';
export type AnnouncementPriority = 'high' | 'low' | 'normal' | 'urgent';
export type AnnouncementDeliveryStatus =
  | 'pending'
  | 'read'
  | 'submitted'
  | string;
export type AnnouncementStatus = 'draft' | 'published';

export interface AnnouncementFormOption {
  label: string;
  value: string;
}

export interface AnnouncementFormField {
  key: string;
  type: AnnouncementFieldType;
  label: string;
  required: boolean;
  placeholder?: string;
  options?: AnnouncementFormOption[];
  must_be_true?: boolean;
}

export type AnnouncementAnswers = Record<string, boolean | string | string[]>;

export interface AnnouncementInfoRaw {
  id: number;
  tenant_id: number;
  scope: string;
  title: string;
  content?: null | string;
  status: AnnouncementStatus | string;
  priority: AnnouncementPriority | string;
  require_response: boolean;
  form_schema?: AnnouncementFormField[] | null;
  published_at?: null | string;
  recipient_count: number;
  response_count: number;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface AnnouncementInfo {
  id: number;
  tenantId: number;
  scope: string;
  title: string;
  content?: null | string;
  status: AnnouncementStatus | string;
  priority: AnnouncementPriority | string;
  requireResponse: boolean;
  formSchema: AnnouncementFormField[];
  publishedAt?: null | string;
  recipientCount: number;
  responseCount: number;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
}

export interface AnnouncementPayload {
  title: string;
  content?: null | string;
  priority: AnnouncementPriority;
  require_response: boolean;
  form_schema: AnnouncementFormField[];
  sort_order?: number;
}

export type AnnouncementUpdatePayload = Partial<AnnouncementPayload>;

export interface AnnouncementListResponse {
  items: AnnouncementInfo[];
  page: number;
  page_size: number;
  total: number;
}

export interface AnnouncementDeliveryRaw {
  id: number;
  tenant_id: number;
  announcement_id: number;
  recipient_type: string;
  recipient_id: number;
  status: string;
  notification_id?: null | number;
  submitted_at?: null | string;
  answers?: null | Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface AnnouncementDelivery {
  id: number;
  tenantId: number;
  announcementId: number;
  recipientType: string;
  recipientId: number;
  status: string;
  notificationId?: null | number;
  submittedAt?: null | string;
  answers?: null | Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface PendingAnnouncementRaw extends AnnouncementInfoRaw {
  delivery_id: number;
}

export interface CurrentAnnouncementRaw extends AnnouncementInfoRaw {
  answers?: null | Record<string, unknown>;
  delivery_id: number;
  delivery_status: AnnouncementDeliveryStatus;
  notification_id?: null | number;
  submitted_at?: null | string;
}

export interface CurrentAnnouncement extends AnnouncementInfo {
  answers?: null | Record<string, unknown>;
  deliveryId: number;
  deliveryStatus: AnnouncementDeliveryStatus;
  notificationId?: null | number;
  submittedAt?: null | string;
}

export type PendingAnnouncement = CurrentAnnouncement;
export type AnnouncementModalItem = CurrentAnnouncement;

export interface AnnouncementSubmitResultRaw {
  id: number;
  announcement_id: number;
  delivery_id: number;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface AnnouncementSubmitResult {
  id: number;
  announcementId: number;
  deliveryId: number;
  status: string;
  createdAt: string;
  updatedAt: string;
}

export interface AnnouncementManagementApi {
  create: (payload: AnnouncementPayload) => Promise<AnnouncementInfo>;
  delete: (id: number) => Promise<void>;
  get: (id: number) => Promise<AnnouncementInfo>;
  getMine?: (id: number) => Promise<CurrentAnnouncement>;
  getResponses: (id: number) => Promise<AnnouncementDelivery[]>;
  list: (params?: Record<string, unknown>) => Promise<AnnouncementListResponse>;
  publish: (id: number) => Promise<AnnouncementInfo>;
  update: (
    id: number,
    payload: AnnouncementUpdatePayload,
  ) => Promise<AnnouncementInfo>;
}

export function transformAnnouncement(
  raw: AnnouncementInfoRaw,
): AnnouncementInfo {
  return {
    id: raw.id,
    tenantId: raw.tenant_id,
    scope: raw.scope,
    title: raw.title,
    content: raw.content,
    status: raw.status,
    priority: raw.priority,
    requireResponse: raw.require_response,
    formSchema: raw.form_schema ?? [],
    publishedAt: raw.published_at,
    recipientCount: raw.recipient_count,
    responseCount: raw.response_count,
    sortOrder: raw.sort_order,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

export function createAnnouncementFormField(
  type: AnnouncementFieldType,
  index: number,
): AnnouncementFormField {
  const key = `${type}_${index + 1}`;
  const field: AnnouncementFormField = {
    key,
    label: '',
    required: true,
    type,
  };

  if (type === 'consent') {
    field.must_be_true = true;
  }

  if (type === 'radio' || type === 'checkbox') {
    field.options = [
      { label: '', value: 'option_1' },
      { label: '', value: 'option_2' },
    ];
  }

  return field;
}

export function createDefaultAnnouncementAnswers(
  schema: AnnouncementFormField[],
): AnnouncementAnswers {
  const answers: AnnouncementAnswers = {};
  for (const field of schema) {
    if (field.type === 'checkbox') {
      answers[field.key] = [];
    } else if (field.type === 'consent') {
      answers[field.key] = false;
    } else {
      answers[field.key] = '';
    }
  }
  return answers;
}

export function normalizeAnnouncementAnswers(
  schema: AnnouncementFormField[],
  rawAnswers?: null | Record<string, unknown>,
): AnnouncementAnswers {
  const answers = createDefaultAnnouncementAnswers(schema);
  if (!rawAnswers) {
    return answers;
  }

  for (const field of schema) {
    const value = rawAnswers[field.key];
    if (field.type === 'consent') {
      if (typeof value === 'boolean') {
        answers[field.key] = value;
      }
      continue;
    }

    if (field.type === 'text' || field.type === 'radio') {
      if (typeof value === 'string') {
        answers[field.key] = value;
      }
      continue;
    }

    if (field.type === 'checkbox' && Array.isArray(value)) {
      answers[field.key] = value.filter(
        (item): item is string => typeof item === 'string',
      );
    }
  }

  return answers;
}

export function validateAnnouncementAnswers(
  schema: AnnouncementFormField[],
  answers: Record<string, unknown>,
): string[] {
  const errors: string[] = [];

  for (const field of schema) {
    const value = answers[field.key];
    const required = Boolean(field.required);

    if (field.type === 'consent') {
      if (value === undefined || value === null) {
        if (required || field.must_be_true)
          errors.push(`${field.key}.required`);
        continue;
      }
      if (typeof value !== 'boolean') {
        errors.push(`${field.key}.boolean`);
        continue;
      }
      if (field.must_be_true && value !== true) {
        errors.push(`${field.key}.must_be_true`);
      }
      continue;
    }

    if (field.type === 'text') {
      if (
        (typeof value !== 'string' || value.trim().length === 0) &&
        required
      ) {
        errors.push(`${field.key}.required`);
      }
      continue;
    }

    const optionValues = new Set(
      (field.options ?? []).map((option) => option.value),
    );

    if (field.type === 'radio') {
      if (typeof value !== 'string' || value.length === 0) {
        if (required) errors.push(`${field.key}.required`);
        continue;
      }
      if (!optionValues.has(value)) {
        errors.push(`${field.key}.invalid_option`);
      }
      continue;
    }

    if (field.type === 'checkbox') {
      if (!Array.isArray(value)) {
        if (required) errors.push(`${field.key}.required`);
        continue;
      }
      if (required && value.length === 0) {
        errors.push(`${field.key}.required`);
      }
      if (
        value.some(
          (item) => typeof item !== 'string' || !optionValues.has(item),
        )
      ) {
        errors.push(`${field.key}.invalid_option`);
      }
    }
  }

  const validKeys = new Set(schema.map((field) => field.key));
  for (const key of Object.keys(answers)) {
    if (!validKeys.has(key)) {
      errors.push(`${key}.unknown`);
    }
  }

  return errors;
}

export function transformDelivery(
  raw: AnnouncementDeliveryRaw,
): AnnouncementDelivery {
  return {
    id: raw.id,
    tenantId: raw.tenant_id,
    announcementId: raw.announcement_id,
    recipientType: raw.recipient_type,
    recipientId: raw.recipient_id,
    status: raw.status,
    notificationId: raw.notification_id,
    submittedAt: raw.submitted_at,
    answers: raw.answers,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

export function transformPendingAnnouncement(
  raw: PendingAnnouncementRaw,
): PendingAnnouncement {
  return {
    ...transformAnnouncement(raw),
    answers: null,
    deliveryId: raw.delivery_id,
    deliveryStatus: 'pending',
    notificationId: null,
    submittedAt: null,
  };
}

export function transformCurrentAnnouncement(
  raw: CurrentAnnouncementRaw,
): CurrentAnnouncement {
  return {
    ...transformAnnouncement(raw),
    answers: raw.answers,
    deliveryId: raw.delivery_id,
    deliveryStatus: raw.delivery_status,
    notificationId: raw.notification_id,
    submittedAt: raw.submitted_at,
  };
}

export function transformSubmitResult(
  raw: AnnouncementSubmitResultRaw,
): AnnouncementSubmitResult {
  return {
    id: raw.id,
    announcementId: raw.announcement_id,
    deliveryId: raw.delivery_id,
    status: raw.status,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}
